README.md
# NAME

WebService::Lob

# VERSION

version 0.0107

# SYNOPSIS

    use WebService::Lob;

    my $lob = WebService::Lob->new( api_key => 'abc123' );

    $lob->get_countries();

# DESCRIPTION

This module provides bindings for the
[Lob](https://www.lob.com/docs) API.

[![Build Status](https://travis-ci.org/aanari/WebService-Lob.svg?branch=master)](https://travis-ci.org/aanari/WebService-Lob)

# METHODS

## new

Instantiates a new WebService::Lob client object.

    my $lob = WebService::Lob->new(
        api_key  => $api_key,
        base_url => $domain,     # optional
        timeout  => $retries,    # optional
        retries  => $retries,    # optional
    );

**Parameters**

- - `api_key`

    _Required_
     

    A valid Lob api key for your account.

- - `base_url`

    _Optional_
     

    The Lob base url to make API calls against.  Defaults to [https://api.lob.com](https://api.lob.com).

- - `timeout`

    _Optional_
     

    The number of seconds to wait per request until timing out.  Defaults to `10`.

- - `retries`

    _Optional_
     

    The number of times to retry requests in cases when Lob returns a 5xx response.  Defaults to `0`.

## get\_states

Returns a list of all US states.

**Request:**

    get_states();

**Response:**

    [{
        name       => 'Alabama',
        short_name => 'AL',
        object     => 'state',
    },
    {
        name       => 'Alaska',
        short_name => 'AK',
        object     => 'state',
    },
    ...
    {
        name       => 'Wisconsin',
        short_name => 'WI',
        object     => 'state',
    },
    {
        name       => 'Wyoming',
        short_name => 'WY',
        object     => 'state',
    }]

## get\_countries

Returns a list of all currently supported countries.

**Request:**

    get_countries();

**Response:**

    [{
        name       => 'United States',
        short_name => 'US',
        object     => 'country',
    },
    {
        name       => 'Afghanistan',
        short_name => 'AF',
        object     => 'country',
    },
    ...
    {
        name       => 'Zambia',
        short_name => 'ZM',
        object     => 'country',
    },
    {
        name       => 'Zimbabwe',
        short_name => 'ZW',
        object     => 'country',
    }]

## verify\_address

Validates an address given.

**Request:**

    verify_address(
        address_line1   => '370 Townsend St',
        address_city    => 'San Francisco',
        address_state   => 'CA',
        address_zip     => '94107',
        address_country => 'US',
    );

**Response:**

    {
        object          => 'address',
        address_line1   => '370 TOWNSEND ST',
        address_line2   => '',
        address_city    => 'SAN FRANCISCO',
        address_state   => 'CA',
        address_zip     => '94107-1607',
        address_country => 'US',
    }

**Exceptions:**

- - `WebService::Lob::Exception::AddressNotFound`

    Address Not Found.

- - `WebService::Lob::Exception::AddressMissingInformation`

    The address you entered was found but more information is needed to match to a specific address.

# BUGS

Please report any bugs or feature requests on the bugtracker website
https://github.com/aanari/WebService-Lob/issues

When submitting a bug or request, please include a test-file or a
patch to an existing test-file that illustrates the bug or desired
feature.

# AUTHOR

Ali Anari <ali@anari.me>

# CONTRIBUTOR

Naveed Massjouni <naveedm9@gmail.com>

# COPYRIGHT AND LICENSE

This software is copyright (c) 2014 by Ali Anari.

This is free software; you can redistribute it and/or modify it under
the same terms as the Perl 5 programming language system itself.


Changes
Revision history for Perl module WebService::Lob

0.0108 2014-12-?? AANARI
    - Added Changes file
    - Added abstract section to pod (=head1 NAME), so module will appear
      right in MetaCPAN etc.
    - Added a SEE ALSO section to pod, with link to lob.com docs page
    - Specified min perl version as 5.8.1 in code and metadata

0.0107 2014-11-07 AANARI
    - Updated to use WebService::Client
    - Switch to Method::Signatures to work with older Perls
    - Fixed dist and weaver for travis

0.0105 2014-08-23 AANARI
    - Remove TrustPod and Pod Coverage for now
    - Fixed pod, bumped version
    - Improved Travis, dzil, and weaver config
    - use aliased exception class names (@ironcamel++)
    - Removed aliased to fix dzil, bumped version

0.0104 2014-08-22 AANARI
    - First release to CPAN



t/Harness.pm
package t::Harness;

use Exporter 'import';
use Memoize;
use WebService::Lob;

memoize 'lob';
sub lob {
    my $api_key = $ENV{LOB_API_KEY};
    WebService::Lob->new(api_key => $api_key) if $api_key;
}

@EXPORT_OK = qw(lob);

1;


t/lib/Harness.pm
package t::lib::Harness;

use Exporter 'import';
use Memoize;
use WebService::Lob;

memoize 'lob';
sub lob {
    my $api_key = $ENV{LOB_API_KEY};
    WebService::Lob->new(api_key => $api_key) if $api_key;
}

@EXPORT_OK = qw(lob);

1;


Makefile.PL

# This file was automatically generated by Dist::Zilla::Plugin::MakeMaker v5.020.
use strict;
use warnings;



use ExtUtils::MakeMaker ;



my %WriteMakefileArgs = (
  "ABSTRACT" => "Lob API Bindings",
  "AUTHOR" => "Ali Anari <ali\@anari.me>",
  "CONFIGURE_REQUIRES" => {
    "ExtUtils::MakeMaker" => 0
  },
  "DISTNAME" => "WebService-Lob",
  "EXE_FILES" => [],
  "LICENSE" => "perl",
  "NAME" => "WebService::Lob",
  "PREREQ_PM" => {
    "Function::Parameters" => 0,
    "Moo" => "1.006000",
    "Throwable::Error" => 0,
    "WebService::BaseClientRole" => "0.0004"
  },
  "TEST_REQUIRES" => {
    "Exporter" => 0,
    "File::Spec" => 0,
    "IO::Handle" => 0,
    "IPC::Open3" => 0,
    "Memoize" => 0,
    "Storable" => 0,
    "Test::Modern" => 0,
    "Test::More" => 0,
    "strict" => 0,
    "warnings" => 0
  },
  "VERSION" => "0.0104",
  "test" => {
    "TESTS" => "t/*.t"
  }
);


my %FallbackPrereqs = (
  "Exporter" => 0,
  "File::Spec" => 0,
  "Function::Parameters" => 0,
  "IO::Handle" => 0,
  "IPC::Open3" => 0,
  "Memoize" => 0,
  "Moo" => "1.006000",
  "Storable" => 0,
  "Test::Modern" => 0,
  "Test::More" => 0,
  "Throwable::Error" => 0,
  "WebService::BaseClientRole" => "0.0004",
  "strict" => 0,
  "warnings" => 0
);


unless ( eval { ExtUtils::MakeMaker->VERSION(6.63_03) } ) {
  delete $WriteMakefileArgs{TEST_REQUIRES};
  delete $WriteMakefileArgs{BUILD_REQUIRES};
  $WriteMakefileArgs{PREREQ_PM} = \%FallbackPrereqs;
}

delete $WriteMakefileArgs{CONFIGURE_REQUIRES}
  unless eval { ExtUtils::MakeMaker->VERSION(6.52) };

WriteMakefile(%WriteMakefileArgs);





lib/WebService/Lob.pm
package WebService::Lob;

use 5.008001;
use Moo;
with 'WebService::Client';

# VERSION

use aliased 'WebService::Lob::Exception::AddressNotFound';
use aliased 'WebService::Lob::Exception::AddressMissingInformation';

use Method::Signatures;

has '+base_url'   => ( default => 'https://api.lob.com/v1' );
has api_key       => ( is => 'ro', required => 1           );
has states_uri    => ( is => 'ro', default => '/states'    );
has countries_uri => ( is => 'ro', default => '/countries' );
has verify_uri    => ( is => 'ro', default => '/verify'    );

method BUILD(...) {
    $self->ua->credentials('api.lob.com:443', '', $self->api_key, '');
}

method get_states {
    my $result = $self->get($self->states_uri);
    return $result->{data} if $result;
}

method get_countries {
    my $result = $self->get($self->countries_uri);
    return $result->{data} if $result;
}

method verify_address(
    Str :$address_line1,
    Str :$address_line2 = '',
    Str :$address_city,
    Str :$address_state,
    Str :$address_zip,
    Str :$address_country
) {
    my $result = $self->post($self->verify_uri, {
        address_line1   => $address_line1,
        address_line2   => $address_line2,
        address_city    => $address_city,
        address_state   => $address_state,
        address_zip     => $address_zip,
        address_country => $address_country,
    });

    AddressNotFound->throw unless $result;
    AddressMissingInformation->throw( message => $result->{message} )
        if $result->{message};
    return $result->{address};
}

=head1 NAME

WebService::Lob - interface to API for lob.com (printing and mailing service)

=head1 SYNOPSIS

    use WebService::Lob;

    my $lob = WebService::Lob->new( api_key => 'abc123' );

    $lob->get_countries();

=head1 DESCRIPTION

This module provides bindings for the
L<Lob|https://www.lob.com/docs> API.

=for markdown [![Build Status](https://travis-ci.org/aanari/WebService-Lob.svg?branch=master)](https://travis-ci.org/aanari/WebService-Lob)

=head1 METHODS

=head2 new

Instantiates a new WebService::Lob client object.

    my $lob = WebService::Lob->new(
        api_key  => $api_key,
        base_url => $domain,     # optional
        timeout  => $retries,    # optional
        retries  => $retries,    # optional
    );

B<Parameters>

=over 4

=item - C<api_key>

I<Required>E<10> E<8>

A valid Lob api key for your account.

=item - C<base_url>

I<Optional>E<10> E<8>

The Lob base url to make API calls against.  Defaults to L<https://api.lob.com|https://api.lob.com>.

=item - C<timeout>

I<Optional>E<10> E<8>

The number of seconds to wait per request until timing out.  Defaults to C<10>.

=item - C<retries>

I<Optional>E<10> E<8>

The number of times to retry requests in cases when Lob returns a 5xx response.  Defaults to C<0>.

=back

=head2 get_states

Returns a list of all US states.

B<Request:>

    get_states();

B<Response:>

    [{
        name       => 'Alabama',
        short_name => 'AL',
        object     => 'state',
    },
    {
        name       => 'Alaska',
        short_name => 'AK',
        object     => 'state',
    },
    ...
    {
        name       => 'Wisconsin',
        short_name => 'WI',
        object     => 'state',
    },
    {
        name       => 'Wyoming',
        short_name => 'WY',
        object     => 'state',
    }]

=head2 get_countries

Returns a list of all currently supported countries.

B<Request:>

    get_countries();

B<Response:>

    [{
        name       => 'United States',
        short_name => 'US',
        object     => 'country',
    },
    {
        name       => 'Afghanistan',
        short_name => 'AF',
        object     => 'country',
    },
    ...
    {
        name       => 'Zambia',
        short_name => 'ZM',
        object     => 'country',
    },
    {
        name       => 'Zimbabwe',
        short_name => 'ZW',
        object     => 'country',
    }]

=head2 verify_address

Validates an address given.

B<Request:>

    verify_address(
        address_line1   => '370 Townsend St',
        address_city    => 'San Francisco',
        address_state   => 'CA',
        address_zip     => '94107',
        address_country => 'US',
    );

B<Response:>

    {
        object          => 'address',
        address_line1   => '370 TOWNSEND ST',
        address_line2   => '',
        address_city    => 'SAN FRANCISCO',
        address_state   => 'CA',
        address_zip     => '94107-1607',
        address_country => 'US',
    }

B<Exceptions:>

=over 4

=item - C<WebService::Lob::Exception::AddressNotFound>

Address Not Found.

=item - C<WebService::Lob::Exception::AddressMissingInformation>

The address you entered was found but more information is needed to match to a specific address.

=back

=head1 SEE ALSO

L<https://lob.com/docs> - the API docs for L<lob.com>

=cut

1;
